lib/ts/nextjs/middleware.ts

```diff
@@ -0,0 +1,204 @@
+        if (hasRequiredCookies) {
+            finalResponse.headers.append("set-cookie", `${tokens.accessToken.cookie as string}`);
+            finalResponse.headers.append("set-cookie", `${tokens.refreshToken.cookie as string}`);
+            finalResponse.headers.append("set-cookie", `${tokens.frontToken.cookie as string}`);
+        }
+        if (hasRequiredHeaders) {
+            finalResponse.headers.append(REFRESH_TOKEN_HEADER_NAME, tokens.refreshToken.header as string);
+            finalResponse.headers.append(ACCESS_TOKEN_HEADER_NAME, tokens.accessToken.header as string);
+        }
+
+        if (tokens.antiCsrfToken.cookie) {
+            finalResponse.headers.append("set-cookie", `${tokens.antiCsrfToken.cookie as string}`);
+        } else if (tokens.antiCsrfToken.header) {
+            finalResponse.headers.append(ANTI_CSRF_TOKEN_HEADER_NAME, tokens.antiCsrfToken.header);
+        }
+        finalResponse.headers.append(
+            "set-cookie",
+            `${REDIRECT_ATTEMPT_COUNT_COOKIE_NAME}=${
+                redirectAttemptNumber + 1
+            }; Path=/; Max-Age=10; HttpOnly; SameSite=Strict`
+        );
+        logDebugMessage("Attached new tokens to response");
+        return finalResponse;
+    } catch (err) {
+        logDebugMessage("Error refreshing session");
+        logDebugMessage(err as unknown as string);
+        return redirectToAuthPage(request);
+    }
+}
+
+export async function revokeSession(config: SuperTokensNextjsConfig): Promise<Response | void> {
+    AppInfo = config.appInfo;
+    if (config.enableDebugLogs) {
+        enableLogging();
+    }
+    const signOutURL = new URL(`${AppInfo.apiBasePath}/signout`, AppInfo.apiDomain);
+    try {
+        const signOutResponse = await fetch(signOutURL, {
+            method: "POST",
+            headers: {
+                "Content-Type": "application/json",
+            },
+            credentials: "include",
+        });
+
+        if (signOutResponse.status !== 200) {
+            return;
+        }
+
+        const response = new Response(null, {});
+        response.headers.set("x-middleware-next", "1");
+        // TODO: Delete only the cookies that are auth related
+        response.headers.delete("set-cookie");
+        response.headers.delete(ACCESS_TOKEN_HEADER_NAME);
+        response.headers.delete(REFRESH_TOKEN_HEADER_NAME);
+        response.headers.delete(FRONT_TOKEN_HEADER_NAME);
+        response.headers.delete(ANTI_CSRF_TOKEN_HEADER_NAME);
+        return response;
+    } catch (err) {
+        logDebugMessage("Error revoking session");
+        logDebugMessage(err as unknown as string);
+        return;
+    }
+}
+
+function redirectToAuthPage(request: Request): Response {
+    const authPagePath = AppInfo.websiteBasePath || "/auth";
+    const redirectUrl = new URL(authPagePath, request.url);
+    redirectUrl.searchParams.set(FORCE_LOGOUT_PATH_PARAM_NAME, "true");
```

Comment:
where is this used?

Reply:
In the following scenarios:
- if the refresh token cannot be found
- if the refresh request fails
- if, due to an unforeseen error, the ssr get session operation ends up in a loop